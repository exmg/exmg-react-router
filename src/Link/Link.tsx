import React, { PureComponent } from 'react';

import { Route } from '../Route';
import { getHistory } from '../utils';

const isModifiedEvent = (event: React.MouseEvent<HTMLAnchorElement>) =>
  !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);

interface Props {
  activeClassName: string;
  back?: boolean;
  children: React.ReactNode;
  className: string;
  exact: boolean;
  onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void;
  relative?: boolean;
  replace: boolean;
  to: string;
}

export default class Link extends PureComponent<Props> {
  static defaultProps = {
    activeClassName: 'is-active',
    className: 'link',
    exact: false,
    replace: false,
    to: '/',
  };

  clickAction = (event: React.MouseEvent<HTMLAnchorElement>) => {
    const { onClick, replace, back } = this.props;
    const slug = event.currentTarget.getAttribute('href');

    if (onClick) {
      onClick(event);
    }

    if (!event.defaultPrevented && event.button === 0 && !isModifiedEvent(event)) {
      event.preventDefault();

      if (back) {
        return getHistory().goBack();
      }

      if (replace) {
        getHistory().replace(slug);
      } else {
        getHistory().push(slug);
      }
    }
  }

  render() {
    const { to, exact, className, activeClassName, children, relative } = this.props;

    return (
      <Route path={ to } exact={ exact } exclude>
        { (match, params, parentPath) => (
            <a
              href={ relative ? `${parentPath}${to}` : to }
              onClick={ this.clickAction }
              className={ `${className} ${match ? activeClassName : ''}` }
            >
              { children }
            </a>
          ) }
      </Route>
    );
  }
}
