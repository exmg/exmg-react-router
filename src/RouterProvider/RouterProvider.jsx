import { Component } from 'react';
import PropTypes from 'prop-types';
import createHistory from 'history/createHashHistory';

export const routerContextTypes = {
	router: PropTypes.shape({
		path: PropTypes.string.isRequired,
		register: PropTypes.func.isRequired,
		history: PropTypes.shape({
			push: PropTypes.func.isRequired,
			replace: PropTypes.func.isRequired,
		}).isRequired,
	}),
};

export default class RouterProvider extends Component {
	static propTypes = {
		children: PropTypes.node,
	};

	static defaultProps = {
		children: null,
	};

	static childContextTypes = routerContextTypes;

	// eslint-disable-next-line react/sort-comp
	history = createHistory();
	routes = [];
	routeMap = {};

	getChildContext() {
		return {
			router: {
				// Pass path to build op child routes
				path: '',
				register: this.register,
				history: this.history,
			},
		};
	}

	componentWillMount() {
		this.unlisten = this.history.listen(this.updateAction);
	}

	componentDidMount() {
		this.updateAction();
	}

	componentWillUnmount() {
		this.unlisten();
	}

	register = (route) => {
		this.routes.push(route);

		const parentPath = route.context.router.path;

		if (!this.routeMap[parentPath]) {
			this.routeMap[parentPath] = [route];
		} else {
			this.routeMap[parentPath].push(route);
		}

		const routeMap = this.routeMap[parentPath];

		this.requestUpdate();

		return () => {
			const index = routeMap.indexOf(route);

			routeMap.splice(index, 1);
		};
	}

	requestUpdate() {
		if (!this.rAF) {
			this.rAF = requestAnimationFrame(this.updateAction);
		}
	}

	updateAction = () => {
		const { pathname } = this.history.location;

		Object.values(this.routeMap).forEach((routes) => {
			let foundMatch = false;
			const nonMatchingRoutes = [];

			routes.forEach((route) => {
				const match = route.match(pathname);

				if (!route.props.notFound) {
					route.update(match);
				}

				if (route.props.exclude) {
					return;
				}

				if (match) {
					foundMatch = true;
				}

				if (route.props.notFound) {
					nonMatchingRoutes.push(route);
				}
			});

			if (nonMatchingRoutes.length) {
				nonMatchingRoutes.forEach(route => route.update(!foundMatch));
			}
		});

		this.rAF = null;
	}

	render() {
		return this.props.children;
	}
}
