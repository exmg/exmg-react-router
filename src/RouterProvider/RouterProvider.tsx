import React, { Component } from 'react';

import { createHistory, HistoryTypes } from '../utils';
import { Route } from 'Route/Route';
import { UnregisterCallback  } from 'history';

export interface Context {
  register: (component: Component) => (() => void);
  parentPath: string;
}

export const Context = React.createContext<Context>({} as Context);

export interface Props {
  children: React.ReactNode;
  type: HistoryTypes;
}

interface RoutesMap {
  [key: string]: Route[];
}

export default class RouterProvider extends Component<Props> {
  static defaultProps = {
    type: 'hash',
  };

  raf?: number;
  routesMap: RoutesMap = {};
  unlisten: UnregisterCallback;
  history = createHistory(this.props.type);

  register: Context['register'] = (routeComponent: Route) => {
    const { parentPath } = routeComponent.context;

    if (!this.routesMap[parentPath]) {
      this.routesMap[parentPath] = [];
    }

    const map = this.routesMap[parentPath];

    map.push(routeComponent);

    this.requestUpdate();

    return () => {
      const index = map.indexOf(routeComponent);

      map.splice(index, 1);

      this.requestUpdate();
    };
  }

  contextProps: Context = {
    register: this.register,
    parentPath: '/',
  };

  componentDidMount() {
    this.unlisten = this.history.listen(this.update);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  requestUpdate() {
    if (!this.raf) {
      this.raf = requestAnimationFrame(this.update);
    }
  }

  update = () => {
    const { pathname } = this.history.location;
    const routeGroups = Object.values(this.routesMap);

    routeGroups.forEach(routes => this.updateGroup(pathname, routes));

    this.raf = null;
  }

  updateGroup(pathname: string, routes: Route[]) {
    let foundMatch = false;
    const nonMatchingRoutes: Route[] = [];

    routes.forEach((route) => {
      const match = route.match(pathname);

      if (!route.props.notFound) {
        route.update(match);
      }

      if (route.props.exclude) {
        return;
      }

      if (match) {
        foundMatch = true;
      }

      if (route.props.notFound) {
        nonMatchingRoutes.push(route);
      }
    });

    if (nonMatchingRoutes.length) {
      const notFoundMatch = foundMatch ? false : { pathname, params: {} };

      nonMatchingRoutes.forEach(route => route.update(notFoundMatch));
    }
  }

  render() {
    return (
      <Context.Provider value={ this.contextProps }>
        { this.props.children }
      </Context.Provider>
    );
  }
}
