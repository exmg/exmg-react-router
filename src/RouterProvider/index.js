export { default as RouterProvider, routerContextTypes } from './RouterProvider';
