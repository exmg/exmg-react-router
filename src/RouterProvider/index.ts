export { default as RouterProvider } from './RouterProvider';
