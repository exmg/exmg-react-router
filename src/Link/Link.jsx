import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { Route } from '../Route';
import { routerContextTypes } from '../RouterProvider/RouterProvider';

const isModifiedEvent = event =>
	!!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);

export default class Link extends PureComponent {
	static propTypes = {
		activeClassName: PropTypes.string,
		children: PropTypes.node,
		className: PropTypes.string,
		exact: PropTypes.bool,
		onClick: PropTypes.func,
		to: PropTypes.string.isRequired,
		replace: PropTypes.bool,
	};

	static defaultProps = {
		activeClassName: 'is-active',
		children: null,
		className: 'link',
		exact: false,
		onClick: null,
		replace: false,
	};

	static contextTypes = routerContextTypes;

	clickAction = (event) => {
		const { to, onClick, replace } = this.props;
		const { history } = this.context.router;

		if (
			!event.defaultPrevented && // onClick prevented default
			event.button === 0 && // ignore right clicks
			!isModifiedEvent(event) // ignore clicks with modifier keys
		) {
			event.preventDefault();

			if (replace) {
				history.replace(to);
			} else {
				history.push(to);
			}
		}

		if (onClick) {
			onClick(event);
		}
	}

	render() {
		const { to, exact, children, className, activeClassName } = this.props;

		return (
			<Route path={ to } exact={ exact } exclude>
				{ match => (
					<a
						href={ `#${to}` }
						onClick={ this.clickAction }
						className={ `${className} ${match ? activeClassName : ''}` }
					>
						{ children }
					</a>
				) }
			</Route>
		);
	}
}
