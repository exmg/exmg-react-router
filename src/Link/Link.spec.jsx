import React from 'react';
import renderer from 'react-test-renderer';

import RouterProvider from '../RouterProvider/RouterProvider';
import Link from './Link';

global.requestAnimationFrame = cb => cb();

test('Link renders children', () => {
	const component = renderer.create(
		<RouterProvider>
			<Link to="/" exact>Home</Link>
			<Link to="/sub">Sub</Link>
		</RouterProvider>,
	);

	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
