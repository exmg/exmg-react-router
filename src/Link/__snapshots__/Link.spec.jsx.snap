// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders children 1`] = `
Array [
  <a
    className="link is-active"
    href="#/"
    onClick={[Function]}
  >
    Home
  </a>,
  <a
    className="link "
    href="#/sub"
    onClick={[Function]}
  >
    Sub
  </a>,
]
`;
